src/Memory/CLState.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import * as BB from 'botbuilder'
+import * as CLM from '@conversationlearner/models'
+import * as Utils from '../Utils'
+import { CLDebug } from '../CLDebug'
+import { EntityState } from './EntityState'
+import { BotState } from './BotState'
+import { InProcessMessageState as MessageState } from './InProcessMessageState'
+import { CLStorage } from '../CLStorage';
+import { BrowserSlotState } from './BrowserSlot';
+
+/**
+ * CLState is a container all states (BotState, EntityState, MessageState) and can be passed around as a single access point.
+ * 
+ * Responsibilities
+ * - Single place to define keys for each of the sub states
+ *  - MessageState is dynamic from conversation
+ *  - Entity and Bot are dynamic from conversation model
+ *  - BrowserSlot is static `BROWSER_SLOTS`
+ * 
+ * - Higher level state operations that span multiple types of state
+ *   - Example: SetApp (which affects BotState, EntityState, and MessageState)
+ */
+export class CLState {
+    private static bbStorage: BB.Storage
+    public readonly turnContext?: BB.TurnContext
+
+    BotState: BotState
+    EntityState: EntityState
+    MessageState: MessageState
+    BrowserSlotState: BrowserSlotState
+
+    private constructor(
+        botState: BotState,
+        entityState: EntityState,
+        messageState: MessageState,
+        browserSlotState: BrowserSlotState,
+        turnContext?: BB.TurnContext) {
+
+        this.BotState = botState
+        this.EntityState = entityState
+        this.MessageState = messageState
+        this.BrowserSlotState = browserSlotState
+
+        this.turnContext = turnContext
+    }
+
+    public static Init(storage?: BB.Storage): void {
+        // If memory storage not defined use disk storage
+        if (!storage) {
+            CLDebug.Log('Storage not defined.  Defaulting to in-memory storage.')
+            storage = new BB.MemoryStorage()
+        }
+
+        CLState.bbStorage = storage
+    }
+
+    public static Get(key: string, modelId: string = ''): CLState {
+        const storage = new CLStorage(CLState.bbStorage)
+
+        const keyPrefix = Utils.getSha256Hash(key)
+        const modelUniqueKeyPrefix = Utils.getSha256Hash(`${modelId}${key}`)
+
+        const botState = new BotState(storage, (datakey) => `${keyPrefix}_${datakey}`)
+        const entityState = new EntityState(storage, () => `${keyPrefix}_BOTMEMORY`)
+        const messageState = new MessageState(storage, () => `${modelUniqueKeyPrefix}_MESSAGE_MUTEX`)
+        const browserSlotState = new BrowserSlotState(storage, () => `BROWSER_SLOTS`)
+
+        return new CLState(botState, entityState, messageState, browserSlotState)
```

Comment:
Knowing what the code was like before, I'd say this is SO much clearer. I want to call out some points because these complicated hierarchies are around various parts of the SDK and I hope this alternate architecture be example of how I would prefer the code to look.

From looking at a single function, you can see that BotState, MessageState, EntityState, etc are children/properties of CLState and you can see that the same `storage` instance is shared across all of them.  This allows people to visualize the hierarchy of program without having to mentally connect references from 4 files